### ISSUE TITLE
Multiple declarations of the same gem are not respected when locking the bundle

### ISSUE THREAD

**Issue by severin** [bundler/bundler#568]
I'm working on a large Rails 2 app and we use Bundler 0.9.26 for dependency management since we use a lot of different gems in different environments. But sometimes these environments overlap. So we have for example the following in our Gemfile:

```
    group(:default) do
      # all shared stuff
    end

    group(:test) do
      # testing stuff
      gem "faker"
    end

    group(:development) do
      # debugger, logging, etc.
      gem "faker"
    end
```

So we basically declare some gems multiple times.

Everything works when using this Gemfile unlocked. However, when I lock the bundle, faker is no longer available in the test environment. The lock file looks like this:

dependencies: 
  faker: 
    group: 
    - :development
    version: = 0.3.1

So basically the last declaration of a gem (development in our case) "wins". I don't know if this is a bug or intended behaviour, but it is at least inconsistent (unlocked it works, locked it does not).
A simple workaround is to declare shared gems in a "supergroup" like:

```
    group(:default) do
      # all shared stuff
    end

    group(:test) do
      # testing stuff
    end

    group(:development) do
      # debugger, logging, etc.
    end

    group(:test, :development) do
      gem "faker"
    end
```
